src/main/java/org/opensearch/neuralsearch/search/HybridTopScoreDocCollector.java

```diff
@@ -113,10 +112,10 @@ public void collect(int doc) throws IOException {
                     }
                     totalHits[i]++;
                     PriorityQueue<ScoreDoc> pq = compoundScores[i];
-                    ScoreDoc topDoc = pq.top();
-                    topDoc.doc = doc + docBase;
-                    topDoc.score = score;
-                    pq.updateTop();
```

Comment:
I have question on the old implementation, even if top of the pq has a score better than current doc we were still updating the top of the doc. this looks to me like a bug. @martin-gaievski how it was working earlier?

Reply:
that PriorityQueue is a [lucene implementation](https://github.com/yintaoxue/read-open-source-code/blob/master/solr-4.10.4/src/org/apache/lucene/util/PriorityQueue.java) of min heap, so top is actually the lowest score. In previous implementation heap got pre-populated with zero score objects, then we iterate over doc scores, put them to heap and push down to the bottom of the heap. This is they call [downHeap in lucene class](https://github.com/yintaoxue/read-open-source-code/blob/master/solr-4.10.4/src/org/apache/lucene/util/PriorityQueue.java#L233) 
Example:
```
- initialization [0, 0, 0, 0, 0] 
- step 1, score 0.5 -> [0, 0, 0, 0, 0.5]
- step 2, score 0.3 -> [0, 0, 0, 0.3, 0.5]
- step 3, score 0.6 -> [0, 0, 0.3, 0.6, 0.5]
- step 4, score 0.2 -> [0, 0.2, 0.3, 0.6, 0.5]
```

When we fill final collection of scores we:
- pop all empty (score 0) elements 
- pop all elements and fill the collection starting from the end going to element with index 0. part of the pop() method is to rebuild the heap so new top element is the min element

Example
```
- we know there were 4 hits, so left only last 4 elements [0, 0.2, 0.3, 0.6, 0.5] -> [0.2, 0.3, 0.6, 0.5]
- [0.2, 0.3, 0.6, 0.5] -> final scores [0.2]  min heap [0.3, 0.6, 0.5]
- [0.3, 0.6, 0.5] -> final scores [0.2, 0.3]  min heap [0.5, 0.6] here we're doing downHeap and pushing min element to the top
- [0.5, 0.6] -> final scores [0.2, 0.3, 0.5]  min heap [0.6]
- [0.6] -> final scores [0.2, 0.3, 0.5, 0.6]  min heap []
```

Reply:
discussed with Navneet offline, there was a bug in old implementation. 

The size of the heap is[ less of equal max number of hits among all sub queries](https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java#L69). This means we're not storing all the scores, but keep only sub set of it in the heap. 
we were updating top element blindly without checking if the element we're adding is less or greater than that element. In case when heap is full that will affect min element. If affected that's alway one element.
With new implementation the bug is not there, this is taken care of by [lessThan](https://github.com/yintaoxue/read-open-source-code/blob/master/solr-4.10.4/src/org/apache/lucene/util/PriorityQueue.java#L151) check. This lessThan is implemented for ScoreDoc in [lucene HitQueue](https://github.com/yintaoxue/read-open-source-code/blob/master/lucene-4.7.2/src/org/apache/lucene/search/HitQueue.java#L76-L82)